### PR TITLE
Add topology data model factory for bindable resources

### PR DESCRIPTION
**Pull request by rottencandy**
**Fixes**: 
https://issues.redhat.com/browse/ODC-6160

**Description**: 

Contribute topology data model factory for bindable services as a dynamic extension.
List of bindable resources is fetched up front because topology extensions require list of resources to be already present.

**Screen shots / Gifs for design review**: 
N/A

**Unit test coverage report**: 

**Browser conformance**: 

- [ ] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

/kind feature

### CODE REVIEW

File: frontend/packages/topology/src/bindable-services/fetch-bindable-services-utils.ts
Comment (sahil143) [inline frontend/packages/topology/src/bindable-services/fetch-bindable-services-utils.ts:45]: we don't need to fetch the bindable resources here but rather watch them. 
create a utility that will return objects of type `WatchK8sResource` and that will be used in topology.
You can create a utility that will be passed to `resources` prop in the DataModelFactory extension.

File: frontend/packages/topology/src/bindable-services/fetch-bindable-services-utils.ts
Comment (sahil143) [inline frontend/packages/topology/src/bindable-services/fetch-bindable-services-utils.ts:48]: ideally, this should also be fetched through useK8sWatchResource hook in topology

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/topology/src/bindable-services/data-transformer.ts
Comment (sahil143) [inline frontend/packages/topology/src/bindable-services/data-transformer.ts:76]: No need to fetch these resources separate,  this function will have two arguments `namespace` and `resources`. resources will contain all the resources fetched through topology.

**Comment by rottencandy** [comment]
/hold

This requires adding support for providing resources through a function instead of supplying the object.
Done in #9726

**Comment by rottencandy** [comment]
/hold cancel

Will be proceeding with fetching bindable service CR on page load, since redesigning dynamic extension requires additional effort.

File: frontend/packages/dev-console/src/components/topology/bindable-services/models.ts
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/bindable-services/models.ts:7]:
```suggestion
  plural: 'bindableservices',
```

File: frontend/packages/dev-console/src/components/topology/bindable-services/models.ts
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/bindable-services/models.ts:8]:
```suggestion
  label: 'Bindable Service',
```

File: frontend/packages/dev-console/src/components/topology/bindable-services/models.ts
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/bindable-services/models.ts:9]:
```suggestion
  labelPlural: 'Bindable Services',
```

File: frontend/packages/dev-console/src/components/topology/index.ts
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/topology/index.ts:18]: This is going to pull a lot of unnecessary dependencies into your main and vendor bundles. Biggest one being `@patternfly/react-topology` which will blow up the analyze CI. Keep these into there own files maybe and import them into `plugin.tsx` directly.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/src/components/topology/index.ts
Comment (rottencandy) [reply frontend/packages/dev-console/src/components/topology/index.ts]: Moved `getBindableServicesTopologyDataModel` to it's own file, and imported directly into plugin file.

File: frontend/packages/dev-console/src/components/topology/bindable-services/models.ts
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/topology/bindable-services/models.ts]: @sahil143 @rottencandy 
The latest localization uses the same PascalCasing as the kind in the label. Have a look at https://github.com/openshift/console/blob/master/frontend/public/models/index.ts

Comment (rottencandy) [reply frontend/packages/dev-console/src/components/topology/bindable-services/models.ts:9]: Reverted `label` & `labelPlural` to PascalCase.

File: frontend/packages/dev-console/src/components/topology/bindable-services/topology-data-model.ts
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/bindable-services/topology-data-model.ts:15]: These utilities are duplicates from the data-transformer file https://github.com/openshift/console/pull/9713/files#diff-49655a575ab5cacc79a9c672170b0ef44fe3cd3628dbbee983cb5e5dd80da1f1

Comment (rottencandy) [reply frontend/packages/dev-console/src/components/topology/bindable-services/topology-data-model.ts:15]: Had forgotten to remove : p

**Review by sahil143 — COMMENTED** [review]
tested locally with dummy CRD and CR
```
kind: CustomResourceDefinition
apiVersion: apiextensions.k8s.io/v1
metadata:
  name: bindableservices.binding.operators.coreos.com
spec:
  group: config.openshift.io
  names:
    plural: bindableservices
    singular: bindableservice
    kind: BindableService
    listKind: BindableServiceList
  scope: Cluster
  versions:
    - name: v1alpha1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          description: >-
            Bindable Services demo. 
             Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
          type: object
          required:
            - spec
          properties:
            apiVersion:
              description: >-
                APIVersion defines the versioned schema of this representation
                of an object. Servers should convert recognized schemas to the
                latest internal value, and may reject unrecognized values. More
                info:
                https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
              type: string
            kind:
              description: >-
                Kind is a string value representing the REST resource this
                object represents. Servers may infer this from the endpoint the
                client submits requests to. Cannot be updated. In CamelCase.
                More info:
                https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
              type: string
            metadata:
              type: object
            spec:
              description: spec holds user settable values for configuration
              type: array
              items:
                type: object
                properties:
                  group:
                    type: string
                  version:
                    type: string
                  kind: 
                    type: string
```
Resource
```
apiVersion: binding.operators.coreos.com/v1alpha1
kind: BindableService
metadata:
  name: bindable-services
spec:
  - group: camel.apache.org
    kind: IntegrationPlatform
    version: v1
```
Bindable resource with external-service label
Kapture 2021-08-20 at 01 33 53

Bindable resource without an external-service label is not visualized
Kapture 2021-08-20 at 01 42 57

UI breaks when clicking on a Bindable node to open the Sidebar but it's being fixed in https://github.com/openshift/console/pull/9841
Screenshot 2021-08-20 at 1 35 33 AM

**Comment by sahil143** [comment]
/label px-approved docs-approved

Added label on behalf of doc, px since epic https://issues.redhat.com/browse/ODC-5722 has ack from docs and px and AC hasn't changed

File: frontend/packages/dev-console/src/components/topology/bindable-services/models.ts
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/bindable-services/models.ts:13]:
```suggestion
  // t('devconsole~BindableServices')
  labelPluralKey: 'devconsole~BindableServices',
```

File: frontend/packages/dev-console/src/components/topology/bindable-services/models.ts
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/bindable-services/models.ts:10]:
```suggestion
  // t('devconsole~BindableService')
  labelKey: 'devconsole~BindableService',
```

File: frontend/packages/dev-console/src/plugin.tsx
Comment (sahil143) [inline frontend/packages/dev-console/src/plugin.tsx:25]: your vendor size is increased to 3.91 MB. import this using absolute path
```suggestion
import { getBindableServiceResources } from './components/topology/bindable-service-resources';
```

File: frontend/packages/dev-console/src/components/topology/index.ts
Comment (sahil143) [inline frontend/packages/dev-console/src/components/topology/index.ts:8]: remove this export from this file

**Review by sahil143 — CHANGES_REQUESTED** [review]

**Review by sahil143 — APPROVED** [review]
/lgtm

**Comment by rohitkrai03** [comment]
/label px-approved 
/label docs-approved

Added label on behalf of doc, px since epic https://issues.redhat.com/browse/ODC-5722 has ack from docs and px and AC hasn't changed

File: frontend/packages/dev-console/locales/en/dev-console.json
Comment (rohitkrai03) [inline frontend/packages/dev-console/locales/en/dev-console.json:1]: Remove this file since you're using `devconsole` for i18n strings now.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/locales/en/dev-console.json
Comment (rottencandy) [reply frontend/packages/dev-console/locales/en/dev-console.json]: Removed.

**Review by rohitkrai03 — APPROVED** [review]
/approve

**Comment by sahil143** [comment]
/lgtm

tested locally, works as expected

**Pull request merged**
